File: backend/src/main/java/com/sunrisedental/service/PatientService.java
package com.sunrisedental.service;

import com.sunrisedental.dto.PagedResponse;
import com.sunrisedental.dto.PatientRequest;
import com.sunrisedental.dto.PatientResponse;

/**
 * PatientService — defines the contract for patient management operations.
 *
 * Implementations must:
 *  - Map between {@link PatientRequest}/{@link PatientResponse} and the entity layer
 *  - Never expose JPA entities to the controller layer
 *  - Apply pagination for all list operations
 */
public interface PatientService {

    /**
     * Creates a new patient record.
     *
     * @param request validated patient data
     * @return the created patient as a response DTO
     */
    PatientResponse create(PatientRequest request);

    /**
     * Returns a paginated, optionally filtered and searched list of patients.
     *
     * @param search   search term for name / contact / ID (empty = no filter)
     * @param status   status filter: "ACTIVE", "INACTIVE", or null/empty for all
     * @param page     0-indexed page number
     * @param size     page size
     * @param sortBy   field name to sort by (e.g. "fullName", "createdAt")
     * @param sortDir  "asc" or "desc"
     * @return paged response wrapper
     */
    PagedResponse<PatientResponse> getAll(
            String search, String status,
            int page, int size,
            String sortBy, String sortDir);

    /**
     * Returns a single patient by ID.
     *
     * @throws com.sunrisedental.exception.ResourceNotFoundException if not found
     */
    PatientResponse getById(Long id);

    /**
     * Updates an existing patient record.
     *
     * @param id      patient ID
     * @param request new data (all fields replaced)
     * @return updated patient as a response DTO
     * @throws com.sunrisedental.exception.ResourceNotFoundException if not found
     */
    PatientResponse update(Long id, PatientRequest request);

    /**
     * Deletes a patient record permanently.
     *
     * @throws com.sunrisedental.exception.ResourceNotFoundException if not found
     */
    void delete(Long id);

    /**
     * Checks whether a patient with the given contact number already exists,
     * optionally excluding a specific patient ID (used for update duplicate check).
     *
     * @param contactNumber the phone number to check
     * @param excludeId     patient ID to exclude (null for new-patient check)
     * @return true if a duplicate exists
     */
    boolean isDuplicateContact(String contactNumber, Long excludeId);

    /**
     * Returns summary counts used by the patient list page stat cards.
     */
    PatientSummary getSummary();

    /**
     * Lightweight summary DTO for patient stat cards.
     */
    record PatientSummary(
            long total,
            long active,
            long inactive,
            long newThisMonth
    ) {}
}

File: frontend/src/services/patientService.js
import api from './api';

/**
 * patientService — all API calls for the Patient Management module.
 *
 * Every method wraps an axios call that uses the shared API instance
 * (base URL, JWT interceptor, 401 redirect are all handled by api.js).
 */
const patientService = {

  /**
   * GET /api/patients
   *
   * @param {Object} params
   * @param {string} [params.search='']      - searches name / contact / ID
   * @param {string} [params.status='']      - ACTIVE | INACTIVE | '' (all)
   * @param {number} [params.page=0]         - 0-indexed
   * @param {number} [params.size=10]
   * @param {string} [params.sortBy='createdAt']
   * @param {string} [params.sortDir='desc']
   */
  getAll({ search = '', status = '', page = 0, size = 10, sortBy = 'createdAt', sortDir = 'desc' } = {}) {
    return api.get('/patients', {
      params: { search, status, page, size, sortBy, sortDir },
    });
  },

  /**
   * GET /api/patients/summary
   * Returns { total, active, inactive, newThisMonth }
   */
  getSummary() {
    return api.get('/patients/summary');
  },

  /**
   * GET /api/patients/check-duplicate?contact=...&excludeId=...
   * Returns { data: boolean }
   */
  checkDuplicate(contact, excludeId = null) {
    const params = { contact };
    if (excludeId != null) params.excludeId = excludeId;
    return api.get('/patients/check-duplicate', { params });
  },

  /**
   * GET /api/patients/:id
   */
  getById(id) {
    return api.get(`/patients/${id}`);
  },

  /**
   * POST /api/patients
   * @param {Object} data - PatientRequest fields
   */
  create(data) {
    return api.post('/patients', data);
  },

  /**
   * PUT /api/patients/:id
   * @param {number} id
   * @param {Object} data - PatientRequest fields
   */
  update(id, data) {
    return api.put(`/patients/${id}`, data);
  },

  /**
   * DELETE /api/patients/:id
   */
  delete(id) {
    return api.delete(`/patients/${id}`);
  },
};

export default patientService;
